<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
        "http://www.w3.org/TR/html4/loose.dtd"
        >
<html>
<font face="Linux Biolinum O, Calibri">
    <head>
        <link rel="shortcut icon" type="image/x-icon" href="../icon.ico">
        <style type="text/css">
            #list { padding-top:0px; padding-bottom:0px }
            #restind { line-height:100%; margin-left:2em; text-indent:-2em }
            #allind { line-height:100%; margin-left:2em; text-indent:0em }
            #second{ line-height:120%;margin-left:5em; text-indent:-2em }
            #normal{ line-height:100%;margin-left:0em; text-indent:0em }
        </style>
        <title>RAxML 8.1.21 command reference</title>
    </head>
    <body><h3>This is the help produced by RAxML version 8.1.21</h3>
    <h4>released by Alexandros Stamatakis on May 21 2015.</h4>
    <p>
        With greatly appreciated code contributions by:<br>
        Andre Aberer (HITS)<br>
        Simon Berger (HITS)<br>
        Alexey Kozlov (HITS)<br>
        Kassian Kobert (HITS)<br>
        David Dao (KIT and HITS)<br>
        Nick Pattengale (Sandia)<br>
        Wayne Pfeiffer (SDSC)<br>
        Akifumi S. Tanabe (NRIFS)<br></p>
    <p>
        Please also consult the <a target="_blank" title="Download the .pdf-document (~200 kB)"
                                   href="http://sco.h-its.org/exelixis/php/countManualNew.php">RAxML-manual</a><br>
        Please report bugs via the RAxML <a href="http://sites.google.com/site/raxmlgui/home">google group</a>!<br>
        Please send us all input files, the exact invocation, details of the HW and operating system, as well as all
        error messages printed to screen.</p>
    <tt>
        <p>raxmlHPC[-SSE3|AVX|-PTHREADS|-PTHREADS-SSE3|-HYBRID|-HYBRID-SSE3|HYBRID-AVX]</p>
        <dd>
            <dl> -s sequenceFileName -n outputFileName -m substitutionModel<br>
                [-a weightFileName] [-A secondaryStructureSubstModel]<br>
                [-b bootstrapRandomNumberSeed] [-B wcCriterionThreshold]<br>
                [-c numberOfCategories] [-C] [-d] [-D]<br>
                [-e likelihoodEpsilon] [-E excludeFileName]<br>
                [-f&nbsp;a|A|b|B|c|C|d|D|e|E|F|g|h|H|i|I|j|J|k|m|n|o|p|P|q|r|R|s|S|t|T|u|v|V|w|W|x|y] [-F]<br>
                [-g groupingFileName] [-G placementThreshold] [-h] [-H]<br>
                [-i initialRearrangementSetting] [-I autoFC|autoMR|autoMRE|autoMRE_IGN]<br>
                [-j] [-J MR|MR_DROP|MRE|STRICT|STRICT_DROP|T_
                <PERCENT>] [-k] [-K]<br>
                    [-L MR|MRE|T_
                    <PERCENT>] [-M]<br>
                        [-o outGroupName1[,outGroupName2[,...]]] [-O]<br>
                        [-p parsimonyRandomSeed] [-P proteinModel]<br>
                        [-q multipleModelFileName] [-r binaryConstraintTree]<br>
                        [-R binaryModelParamFile] [-S secondaryStructureFile] [-t userStartingTree]<br>
                        [-T numberOfThreads] [-u] [-U] [-v] [-V] [-w outputDirectory] [-W slidingWindowSize]<br>
                        [-x rapidBootstrapRandomNumberSeed] [-X] [-y] [-Y
                        quartetGroupingFileName|ancestralSequenceCandidatesFileName]<br>
                        [-z multipleTreesFile] [-#|-N numberOfRuns|autoFC|autoMR|autoMRE|autoMRE_IGN]<br>
                        [--mesquite][--silent][--no-seq-check][--no-bfgs]<br>
                        [--asc-corr=stamatakis|felsenstein|lewis]<br>
                        [--flag-check][--auto-prot=ml|bic|aic|aicc]<br>
                        [--epa-keep-placements=number][--epa-accumulated-threshold=threshold]<br>
                        [--epa-prob-threshold=threshold]<br>
                        [--JC69][--K80][--asc-miss=fraction]<br>
            </dl>
        </dd>
    </tt>
    <div id="restind">
        <p><tt>-a</tt><br>
            Specify a column weight file name to assign individual weights to each column of the alignment. Those
            weights must be integers separated by any type and number of whitespaces whithin a separate file, see
            file &ldquo;example_weights&rdquo; for an example.</p>
        <p><tt>-A</tt><br>
            Specify one of the secondary structure substitution models implemented in RAxML. The same nomenclature as in
            the <a href="http://www.bioinf.manchester.ac.uk/resources/phase/">PHASE</a> <a
                    href="http://intranet.cs.man.ac.uk/ai/Software/phase/manual/node97.html">manual</a> is used,
            available models: S6A, S6B, S6C, S6D, S6E, S7A, S7B, S7C, S7D, S7E, S7F, S16, S16A, S16B<br>
            DEFAULT: 16-state GTR model (S16)</p>
        <p><tt>-b</tt><br>
            Specify an integer number (random seed) and turn on bootstrapping<br>
            DEFAULT: OFF</p>
        <p><tt>-B</tt><br>
            specify a floating point number between 0.0 and 1.0 that will be used as cutoff threshold for the MR-based
            bootstopping criteria. The recommended setting is 0.03.<br>
            DEFAULT: 0.03 (recommended empirically determined setting)</p>
        <p><tt>-c</tt><br>
            Specify number of distinct rate categories for RAxML the when model of rate heterogeneity is set to CAT.<br>
            Individual per-site rates are categorized into <tt>numberOfCategories</tt> rate categories to accelerate
            computations. <br>
            DEFAULT: 25</p>
        <p><tt>-C</tt><br>
            Enable verbose output for the &ldquo;<tt>-L</tt>&rdquo; and &ldquo;<tt>-f&nbsp;i</tt>&rdquo; options. This
            will produce more, as well as more verbose output files.<br>
            DEFAULT: OFF</p>
        <p><tt>-d</tt><br>
            start ML optimization from random starting tree <br>
            DEFAULT: OFF</p>
        <p><tt>-D</tt><br>
            ML search convergence criterion. This will break off ML searches if the relative Robinson-Foulds distance
            between the trees obtained from two consecutive lazy SPR cycles is smaller or equal to 1&nbsp;%. Usage
            recommended for very large datasets in terms of taxa.<br>
            On trees with more than 500 taxa this will yield execution time improvements of approximately 50&nbsp;%
            while yielding only slightly worse trees.<br>
            DEFAULT: OFF</p>
        <p><tt>-e</tt><br>
            set model optimization precision in log likelihood units for final optimization of tree topology.<br></p>
    </div>
    <div id="allind" -->
        <table frame="void">
            <tr>
                <td>DEFAULT:</td>
                <td align="right">0.1</td>
                <td>for models not using proportion of invariant sites estimate</td>
            </tr>
            <tr>
                <td></td>
                <td align="right">0.001</td>
                <td>for models using proportion of invariant sites estimate</td>
            </tr>
        </table>
    </div>
    <div id="restind">
        <p><tt>-E</tt><br>
            specify an exclude file name, that contains a specification of alignment positions you wish to exclude.<br>
            Format is similar to Nexus, the file shall contain entries like &ldquo;<tt>100-200 300-400</tt>&rdquo;, to
            exclude a single column write, e.&nbsp;g., &ldquo;<tt>100-100</tt>&rdquo;, if you use a mixed model, an
            appropriatly adapted model file will be written.</p>
        <p><tt>-f</tt><br>
            select algorithm:</p></div>
    <div id="second">
        <p> &ldquo;<tt>-f&nbsp;a</tt>&rdquo;: <br>
            rapid Bootstrap analysis and search for best-scoring ML tree in one program run.</p>
        <p> &ldquo;<tt>-f&nbsp;A</tt>&rdquo;: <br>
            compute marginal ancestral states on a ROOTED reference tree provided with &ldquo;<tt>-t</tt>&rdquo;</p>
        <p> &ldquo;<tt>-f&nbsp;b</tt>&rdquo;: <br>
            draw bipartition information on a tree provided with &ldquo;<tt>-t</tt>&rdquo; based on multiple trees (e.&nbsp;g.,
            from a bootstrap) in a file specifed by &ldquo;<tt>-z</tt>&rdquo;</p>
        <p> &ldquo;<tt>-f&nbsp;B</tt>&rdquo;: <br>
            optimize br-len scaler and other model parameters (GTR, alpha, etc.) on a tree provided
            with &ldquo;<tt>-t</tt>&rdquo;.<br>
            The tree needs to contain branch lengths. The branch lengths will not be optimized, just scaled by a single
            common value.</p>
        <p> &ldquo;<tt>-f&nbsp;c</tt>&rdquo;: <br>
            check if the alignment can be properly read by RAxML</p>
        <p> &ldquo;<tt>-f&nbsp;C</tt>&rdquo;: <br>
            ancestral sequence test for Jiajie, users will also need to provide a list of taxon names
            via &ldquo;<tt>-Y</tt>&rdquo; separated by whitespaces.</p>
        <p> &ldquo;<tt>-f&nbsp;d</tt>&rdquo;: <br>
            new rapid hill-climbing<br>
            DEFAULT: ON</p>
        <p> &ldquo;<tt>-f&nbsp;D</tt>&rdquo;: <br>
            rapid hill-climbing with RELL bootstraps.</p>
        <p> &ldquo;<tt>-f&nbsp;e</tt>&rdquo;: <br>
            optimize model + branch lengths for given input tree under <tt>GAMMA</tt>/<tt>GAMMAI</tt> only</p>
        <p> &ldquo;<tt>-f&nbsp;E</tt>&rdquo;: <br>
            execute very fast experimental tree search, at present only for testing.</p>
        <p> &ldquo;<tt>-f&nbsp;F</tt>&rdquo;: <br>
            execute fast experimental tree search, at present only for testing.</p>
        <p> &ldquo;<tt>-f&nbsp;g</tt>&rdquo;: <br>
            compute per site log Likelihoods for one ore more trees passed via &ldquo;<tt>-z</tt>&rdquo; and write them
            to a file that can be read by <a href="http://www.is.titech.ac.jp/~shimo/prog/consel/">CONSEL</a><br>
            The model parameters will be estimated on the first tree only!</p>
        <p> &ldquo;<tt>-f&nbsp;G</tt>&rdquo;: <br>
            compute per site log Likelihoods for one ore more trees passed via &ldquo;<tt>-z</tt>&rdquo; and write them
            to a file that can be read by <a href="http://www.is.titech.ac.jp/~shimo/prog/consel/">CONSEL</a>.<br>
            The model parameters will be re-estimated for each tree.</p>
        <p> &ldquo;<tt>-f&nbsp;h</tt>&rdquo;: <br>
            compute log likelihood test (SH-test) between best tree passed via &ldquo;<tt>-t</tt>&rdquo; and a bunch of
            other trees passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            The model parameters will be estimated on the first tree only!</p>
        <p> &ldquo;<tt>-f&nbsp;H</tt>&rdquo;: <br>
            compute log likelihood test (SH-test) between best tree passed via &ldquo;<tt>-t</tt>&rdquo; and a bunch of
            other trees passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            The model parameters will be re-estimated for each tree.</p>
        <p> &ldquo;<tt>-f&nbsp;i</tt>&rdquo;: <br>
            calculate IC and TC scores (<a href="http://dx.doi.org/10.1038/nature12130">Salichos and Rokas 2014</a>) on
            a tree provided with &ldquo;<tt>-t</tt>&rdquo; based on multiple trees (e.&nbsp;g., from a bootstrap) in a
            file specified by &ldquo;<tt>-z</tt>&rdquo;.</p>
        <p> &ldquo;<tt>-f&nbsp;I</tt>&rdquo;: <br>
            a simple tree rooting algorithm for unrooted trees.<br>
            It roots the tree by rooting it at the branch that best balances the subtree lengths (sum over branches in
            the subtrees) of the left and right subtree.<br>
            A branch with an optimal balance does not always exist!<br>
            You need to specify the tree you want to root via &ldquo;<tt>-t</tt>&rdquo;.</p>
        <p> &ldquo;<tt>-f&nbsp;j</tt>&rdquo;: <br>
            generate a bunch of bootstrapped alignment files from an original alignemnt file.<br>
            You need to specify a seed with &ldquo;<tt>-b</tt>&rdquo; and the number of replicates
            with &ldquo;<tt>-#</tt>&rdquo; </p>
        <p> &ldquo;<tt>-f&nbsp;J</tt>&rdquo;: <br>
            Compute SH-like support values on a given tree passed via &ldquo;<tt>-t</tt>&rdquo;.</p>
        <p> &ldquo;<tt>-f&nbsp;k</tt>&rdquo;: <br>
            Fix long branch lengths in partitioned data sets with missing data using the branch length stealing
            algorithm.<br>
            This option only works in conjunction with &ldquo;<tt>-t</tt>&rdquo;, &ldquo;<tt>-M</tt>&rdquo;, and &ldquo;<tt>-q</tt>&rdquo;.<br>
            It will print out a tree with shorter branch lengths, but having the same likelihood score.</p>
        <p> &ldquo;<tt>-f&nbsp;m</tt>&rdquo;: <br>
            compare bipartitions between two bunches of trees passed via &ldquo;<tt>-t</tt>&rdquo;
            and &ldquo;<tt>-z</tt>&rdquo; respectively. This will return the Pearson correlation between all
            bipartitions found in the two tree files. A file called <tt>RAxML_bipartitionFrequencies.outputFileName</tt>
            will be printed that contains the pair-wise bipartition frequencies of the two sets.</p>
        <p> &ldquo;<tt>-f&nbsp;n</tt>&rdquo;: <br>
            compute the log likelihood score of all trees contained in a tree file provided by &ldquo;<tt>-z</tt>&rdquo;
            under <tt>GAMMA</tt> or <tt>GAMMA+P-Invar</tt>.<br>
            The model parameters will be estimated on the first tree only!</p>
        <p> &ldquo;<tt>-f&nbsp;N</tt>&rdquo;: <br>
            compute the log likelihood score of all trees contained in a tree file provided by &ldquo;<tt>-z</tt>&rdquo;
            under <tt>GAMMA</tt> or <tt>GAMMA+P-Invar</tt>.<br>
            The model parameters will be re-estimated for each tree.</p>
        <p> &ldquo;<tt>-f&nbsp;o</tt>&rdquo;: <br>
            old and slower rapid hill-climbing without heuristic cutoff</p>
        <p> &ldquo;<tt>-f&nbsp;p</tt>&rdquo;: <br>
            perform pure stepwise MP addition of new sequences to an incomplete starting tree and exit.</p>
        <p> &ldquo;<tt>-f&nbsp;P</tt>&rdquo;: <br>
            perform a phylogenetic placement of sub trees specified in a file passed via &ldquo;<tt>-z</tt>&rdquo; into
            a given reference tree in which these subtrees are contained that is passed via &ldquo;<tt>-t</tt>&rdquo;
            using the evolutionary placement algorithm.</p>
        <p> &ldquo;<tt>-f&nbsp;q</tt>&rdquo;: <br>
            fast quartet calculator.</p>
        <p> &ldquo;<tt>-f&nbsp;r</tt>&rdquo;: <br>
            compute pairwise Robinson-Foulds (RF) distances between all pairs of trees in a tree file passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            If the trees have node labales represented as integer support values the program will also compute two
            flavors of the weighted Robinson-Foulds (WRF) distance.</p>
        <p> &ldquo;<tt>-f&nbsp;R</tt>&rdquo;: <br>
            compute all pairwise Robinson-Foulds (RF) distances between a large reference tree passed
            via &ldquo;<tt>-t</tt>&rdquo; and many smaller trees (that must have a subset of the taxa of the large tree)
            passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            This option is intended for checking the plausibility of very large phylogenies that can not be inspected
            visually any more.</p>
        <p> &ldquo;<tt>-f&nbsp;s</tt>&rdquo;: <br>
            split up a multi-gene partitioned alignment into the respective subalignments.</p>
        <p> &ldquo;<tt>-f&nbsp;S</tt>&rdquo;: <br>
            compute site-specific placement bias using a leave one out test inspired by the evolutionary placement
            algorithm.</p>
        <p> &ldquo;<tt>-f&nbsp;t</tt>&rdquo;: <br>
            do randomized tree searches on one fixed starting tree.</p>
        <p> &ldquo;<tt>-f&nbsp;T</tt>&rdquo;: <br>
            do final thorough optimization of ML tree from rapid bootstrap search in stand-alone mode.</p>
        <p> &ldquo;<tt>-f&nbsp;u</tt>&rdquo;: <br>
            execute morphological weight calibration using maximum likelihood, this will return a weight vector.<br>
            You need to provide a morphological alignment and a reference tree via &ldquo;<tt>-t</tt>&rdquo; </p>
        <p> &ldquo;<tt>-f&nbsp;v</tt>&rdquo;: <br>
            classify a bunch of environmental sequences into a reference tree using thorough read insertions.<br>
            You will need to start RAxML with a non-comprehensive reference tree and an alignment containing all
            sequences (reference + query).</p>
        <p> &ldquo;<tt>-f&nbsp;V</tt>&rdquo;: <br>
            classify a bunch of environmental sequences into a reference tree using thorough read insertions.<br>
            You will need to start RAxML with a non-comprehensive reference tree and an alignment containing all
            sequences (reference + query).<br>
            WARNING: this is a test implementation for more efficient handling of multi-gene/whole-genome datasets!</p>
        <p> &ldquo;<tt>-f&nbsp;w</tt>&rdquo;: <br>
            Compute ELW test on a bunch of trees passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            The model parameters will be estimated on the first tree only!</p>
        <p> &ldquo;<tt>-f&nbsp;W</tt>&rdquo;: <br>
            Compute ELW test on a bunch of trees passed via &ldquo;<tt>-z</tt>&rdquo;.<br>
            The model parameters will be re-estimated for each tree.</p>
        <p> &ldquo;<tt>-f&nbsp;x</tt>&rdquo;: <br>
            compute pair-wise ML distances, ML model parameters will be estimated on an MP starting tree or a
            user-defined tree passed via &ldquo;<tt>-t</tt>&rdquo;, only allowed for GAMMA-based models of rate
            heterogeneity</p>
        <p> &ldquo;<tt>-f&nbsp;y</tt>&rdquo;: <br>
            classify a bunch of environmental sequences into a reference tree using parsimony.<br>
            You will need to start RAxML with a non-comprehensive reference tree and an alignment containing all
            sequences (reference + query).</p></div>
    <div id="restind">
        <p><br>
            DEFAULT for &ldquo;<tt>-f</tt>&rdquo;: new rapid hill climbing</p>
        <p><tt>-F</tt><br>
            enable ML tree searches under CAT model for very large trees without switching to GAMMA in the end (saves
            memory).<br>
            This option can also be used with the GAMMA models in order to avoid the thorough optimization of the
            best-scoring ML tree in the end.<br>
            DEFAULT: OFF</p>
        <p><tt>-g</tt><br>
            specify the file name of a multifurcating constraint tree.<br>
            This tree does not need to be comprehensive, i.&nbsp;e. must not contain all taxa.</p>
        <p><tt>-G</tt><br>
            enable the ML-based evolutionary placement algorithm heuristics by specifiyng a threshold value (fraction of
            insertion branches to be evaluated using slow insertions under ML).</p>
        <p><tt>-h</tt><br>Display this help message.</p>
        <p><tt>-H</tt><br>Disable pattern compression.<br>
            DEFAULT: ON</p>
        <p><tt>-i</tt><br>
            Initial rearrangement setting for the subsequent application of topological changes phase.</p>
        <p><tt>-I</tt><br>
            a posteriori bootstopping analysis. Use:</p></div>
    <div id="second">
        <p>&ldquo;<tt>-I autoFC</tt>&rdquo;<br> for the frequency-based criterion</p>
        <p>&ldquo;<tt>-I autoMR</tt>&rdquo;<br> for the majority-rule consensus tree criterion</p>
        <p>&ldquo;<tt>-I autoMRE</tt>&rdquo;<br> for the extended majority-rule consensus tree criterion</p>
        <p>&ldquo;<tt>-I autoMRE_IGN</tt>&rdquo;<br> for metrics similar to MRE, but include bipartitions under the
            threshold whether they are compatible or not. This emulates MRE but is faster to compute.</p></div>
    <p style="text-indent:2em">You also need to pass a tree file containing several bootstrap replicates via &ldquo;<tt>-z</tt>&rdquo;
    </p>
    <div id="restind">
        <p><tt>-j</tt><br>
            Specifies that intermediate tree files shall be written to file during the standard ML and BS tree searches.<br>
            DEFAULT: OFF</p>
        <p><tt>-J</tt><br>
            Compute majority rule consensus tree with &ldquo;<tt>-J&nbsp;MR</tt>&rdquo; or extended majority rule
            consensus tree with &ldquo;<tt>-J&nbsp;MRE</tt>&rdquo; or strict consensus tree with &ldquo;<tt>-J&nbsp;STRICT</tt>&rdquo;.
            For a custom consensus threshold <tt>>= 50%</tt>, specify &ldquo;<tt>T_
                <NUM>
            </tt>&rdquo;, where <tt>100 >= NUM >= 50</tt>.<br>
            Options &ldquo;<tt>-J STRICT_DROP</tt>&rdquo; and &ldquo;<tt>-J MR_DROP</tt>&rdquo; will execute an
            algorithm that identifies dropsets which contain rogue taxa as proposed by Pattengale et al. in the
            paper &ldquo;Uncovering hidden phylogenetic consensus&rdquo;.<br>
            You will also need to provide a tree file containing several UNROOTED trees via &ldquo;<tt>-z</tt>&rdquo;.
        </p>
        <p><tt>-k</tt><br>
            Specifies that bootstrapped trees should be printed with branch lengths.<br>
            The bootstraps will run a bit longer, because model parameters will be optimized at the end of each run
            under GAMMA or GAMMA+P-Invar respectively.<br>
            DEFAULT: OFF</p>
        <p><tt>-K</tt><br>
            Specify one of the multi-state substitution models (max 32 states) implemented in RAxML.<br>
            Available models are: ORDERED, MK, GTR<br>
            DEFAULT: GTR model </p>
        <p><tt>-L</tt><br>
            Compute consensus trees labelled by IC supports and the overall TC value as proposed in Salichos and Rokas
            2013.<br>
            Compute a majority rule consensus tree with &ldquo;<tt>-L&nbsp;MR</tt>&rdquo; or an extended majority rule
            consensus tree with &ldquo;<tt>-L&nbsp;MRE</tt>&rdquo;.<br>
            For a custom consensus threshold <tt>>= 50%</tt>, specify &ldquo;<tt>-L&nbsp;T_
                <NUM>
            </tt>&rdquo;, where <tt>100 >= NUM >= 50</tt>.<br>
            You will of course also need to provide a tree file containing several UNROOTED trees via &ldquo;<tt>-z</tt>&rdquo;!
        </p>
        <p><tt>-m</tt><br>
            Model of Binary (Morphological), Nucleotide, Multi-State, or Amino Acid Substitution: <br><br>
            BINARY:</p></div>
    <div id="second">
        <p>&ldquo;<tt>-m BINCAT[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>BINGAMMA[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m BINCATI[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>BINGAMMAI[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m ASC_BINCAT[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>BINGAMMA[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.</p>
        <p>&ldquo;<tt>-m BINGAMMA[X]</tt>&rdquo;:<br>
            GAMMA model of rate heterogeneity (alpha parameter will be estimated).<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m BINGAMMAI[X]</tt>&rdquo;:<br>
            Same as <tt>BINGAMMA[X]</tt>, but with estimate of proportion of invariable sites.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m ASC_BINGAMMA[X]</tt>&rdquo;:<br>
            GAMMA model of rate heterogeneity (alpha parameter will be estimated).<7br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
    </div>
    <p style="text-indent:2em">NUCLEOTIDES:</p>
    <div id="second">
        <p>&ldquo;<tt>-m GTRCAT[X]</tt>&rdquo;:<br>
            GTR + Optimization of substitution rates + Optimization of site-specific evolutionary rates which are
            categorized into <tt>numberOfCategories</tt> distinct rate categories for greater computational efficiency.
            Final tree might be evaluated under <tt>GTRGAMMA[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m GTRCATI[X]</tt>&rdquo;:<br>
            GTR + Optimization of substitution rates + Optimization of site-specific evolutionary rates which are
            categorized into <tt>numberOfCategories</tt> distinct rate categories for greater computational efficiency.
            Final tree might be evaluated under <tt>GTRGAMMAI[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m ASC_GTRCAT[X]</tt>&rdquo;:<br>
            GTR + Optimization of substitution rates + Optimization of site-specific evolutionary rates which are
            categorized into numberOfCategories distinct rate categories for greater computational efficiency. Final
            tree might be evaluated under <tt>GTRGAMMA[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.</p>
        <p>&ldquo;<tt>-m GTRGAMMA[X]</tt>&rdquo;:<br>
            GTR + Optimization of substitution rates + GAMMA model of rate heterogeneity (alpha parameter will be
            estimated).<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m GTRGAMMAI[X]</tt>&rdquo;:<br>
            Same as <tt>GTRGAMMA[X]</tt>, but with estimate of proportion of invariable sites.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m ASC_GTRGAMMA[X]</tt>&rdquo;:<br>
            GTR + Optimization of substitution rates + GAMMA model of rate heterogeneity (alpha parameter will be
            estimated).<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
    </div>
    <p style="text-indent:2em"> MULTI-STATE:</p>
    <div id="second">
        <p>&ldquo;<tt>-m MULTICAT[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>MULTIGAMMA[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m MULTICATI[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>MULTIGAMMAI[X]</tt>, depending on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies. </p>
        <p>&ldquo;<tt>-m ASC_MULTICAT[X]</tt>&rdquo;:<br>
            Optimization of site-specific evolutionary rates which are categorized into <tt>numberOfCategories</tt>
            distinct rate categories for greater computational efficiency. Final tree might be evaluated automatically
            under <tt>MULTIGAMMA[X]</tt>, depending on the tree search option.<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m MULTIGAMMA[X]</tt>&rdquo;:<br>
            GAMMA model of rate heterogeneity (alpha parameter will be estimated).<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m MULTIGAMMAI[X]</tt>&rdquo;:<br>
            Same as <tt>MULTIGAMMA[X]</tt>, but with estimate of proportion of invariable sites.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m ASC_MULTIGAMMA[X]</tt>&rdquo;:<br>
            GAMMA model of rate heterogeneity (alpha parameter will be estimated).
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base
            frequencies.<br><br>
            You can use up to 32 distinct character states to encode multi-state regions, they must be used in the
            following order:<br>
            <tt>0, 1, 2, 3, 4, 5, 6, 7, 8, 9, A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V</tt><br>
            i.&nbsp;e., if you have 6 distinct character states you would use <tt>0, 1, 2, 3, 4, 5</tt> to encode these.<br>
            The substitution model for the multi-state regions can be selected via the &ldquo;<tt>-K</tt>&rdquo; option
        </p></div>
    <p style="text-indent:2em">AMINO ACIDS:</p>
    <div id="second">
        <p>&ldquo;<tt>-m PROTCATmatrixName[F|X]</tt>&rdquo;:<br>
            specified AA matrix + Optimization of substitution rates + Optimization of site-specific evolutionary rates
            which are categorized into <tt>numberOfCategories</tt> distinct rate categories for greater computational
            efficiency. Final tree might be evaluated automatically under <tt>PROTGAMMAmatrixName[F|X]</tt>, depending
            on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies. </p>
        <p> &ldquo;<tt>-m PROTCATImatrixName[F|X]</tt>&rdquo;:<br>
            specified AA matrix + Optimization of substitution rates + Optimization of site-specific evolutionary rates
            which are categorized into <tt>numberOfCategories</tt> distinct rate categories for greater computational
            efficiency. Final tree might be evaluated automatically under <tt>PROTGAMMAImatrixName[F|X]</tt>, depending
            on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p>&ldquo;<tt>-m&nbsp;ASC_PROTCATmatrixName[F|X]</tt>&rdquo;:<br>
            specified AA matrix + Optimization of substitution rates + Optimization of site-specific evolutionary rates
            which are categorized into <tt>numberOfCategories</tt> distinct rate categories for greater computational
            efficiency. Final tree might be evaluated automatically under <tt>PROTGAMMAmatrixName[F|X]</tt>, depending
            on the tree search option.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.</p>
        <p> &ldquo;<tt>-m&nbsp;PROTGAMMAmatrixName[F|X]</tt>&rdquo;:<br>
            specified AA matrix + Optimization of substitution rates + GAMMA model of rate heterogeneity (alpha
            parameter will be estimated).<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.</p>
        <p> &ldquo;<tt>-m PROTGAMMAImatrixName[F|X]</tt>&rdquo;:<br>
            Same as <tt>PROTGAMMAmatrixName[F|X]</tt>, but with estimate of proportion of invariable sites.<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies. </p>
        <p> &ldquo;<tt>-m&nbsp;ASC_PROTGAMMAmatrixName[F|X]</tt>&rdquo;:<br>
            specified AA matrix + Optimization of substitution rates + GAMMA model of rate heterogeneity (alpha
            parameter will be estimated).<br>
            With the optional &ldquo;<tt>X</tt>&rdquo; appendix you can specify a ML estimate of base frequencies.<br>
            The <tt>ASC</tt> prefix will correct the likelihood for ascertainment bias.</p>
    </div>
    <p style="margin-left:2em">Available AA substitution models: <tt>DAYHOFF</tt>,<tt> DCMUT</tt>,<tt> JTT</tt>,<tt>
        MTREV</tt>,<tt> WAG</tt>,<tt> RTREV</tt>,<tt> CPREV</tt>,<tt> VT</tt>,<tt> BLOSUM62</tt>,<tt> MTMAM</tt>,<tt>
        LG</tt>,<tt> MTART</tt>,<tt> MTZOA</tt>,<tt> PMB</tt>,<tt> HIVB</tt>,<tt> HIVW</tt>,<tt> JTTDCMUT</tt>,<tt>
        FLU</tt>,<tt> STMTREV</tt>,<tt> DUMMY</tt>, <tt> DUMMY2</tt>,<tt> AUTO</tt>,<tt> LG4M</tt>,<tt> LG4X</tt>,<tt>
        PROT_FILE</tt>, <tt> GTR_UNLINKED</tt>,<tt> GTR</tt><br>
        With the optional &ldquo;<tt>F</tt>&rdquo; appendix you can specify if you want to use empirical base
        frequencies.<br>
        <tt>AUTOF</tt> and <tt>AUTOX</tt> are not supported any more, if you specify <tt>AUTO</tt> it will test prot
        subst. models with and without empirical
        base frequencies now!<br>
        Please note that for partitioned models you can in addition specify the per-gene AA model in the partition file
        (see <a href="http://sco.h-its.org/exelixis/php/countManualNew.php">manual</a> for details). Also note that if
        you estimate AA GTR parameters on a partitioned dataset, they will be linked (estimated jointly) across all
        partitions to avoid over-parametrization.</p>
    <div id="restind">
        <p><tt>-M</tt><br>Switch on estimation of individual per-partition branch lengths. Only has effect when used in
            combination with &ldquo;<tt>-q</tt>&rdquo;<br>
            Branch lengths for individual partitions will be printed to separate files.<br>
            A weighted average of the branch lengths is computed by using the respective partition lengths.<br>
            DEFAULT: OFF</p>
        <p><tt>-n</tt><br>Specifies the name of the output file.</p>
        <p><tt>-o</tt><br>Specify the name of a single outgroup or a comma-separated list of outgroups,
            e.&nbsp;g. &ldquo;<tt>-o&nbsp;Rat</tt>&rdquo; or &ldquo;<tt>-o&nbsp;Rat,Mouse</tt>&rdquo;, in case that
            multiple outgroups are not monophyletic the first name in the list will be selected as outgroup, don't leave
            spaces between taxon names!</p>
        <p><tt>-O</tt><br>Disable check for completely undetermined sequence in alignment.<br>
            The program will not exit with an error message when "-O" is specified.<br>
            DEFAULT: check enabled</p>
        <p><tt>-p</tt><br>Specify a random number seed for the parsimony inferences. This allows you to reproduce your
            results and will help me debug the program.</p>
        <p><tt>-P</tt><br>Specify the file name of a user-defined AA (Protein) substitution model. This file must
            contain 420 entries, the first 400 being the AA substitution rates (this must be a symmetric matrix) and the
            last 20 are the empirical base frequencies.</p>
        <p><tt>-q</tt><br>Specify the file name which contains the assignment of models to alignment partitions for
            multiple models of substitution. For the syntax of this file please consult the <a
                    href="http://sco.h-its.org/exelixis/php/countManualNew.php">manual</a>.</p>
        <p><tt>-r</tt><br>Specify the file name of a binary constraint tree.<br>
            This tree does not need to be comprehensive, i.&nbsp;e. must not contain all taxa.</p>
        <p><tt>-R</tt><br>Specify the file name of a binary model parameter file that has previously been generated with
            RAxML using the &ldquo;<tt>-f&nbsp;e</tt>&rdquo; tree evaluation option. The file name should be: <tt>RAxML_binaryModelParameters.runID</tt>.
        </p>
        <p><tt>-s</tt><br>Specify the name of the alignment data file in PHYLIP format.</p>
        <p><tt>-S</tt><br>Specify the name of a secondary structure file. The file can contain &ldquo;<tt>.</tt>&rdquo;
            for alignment columns that do not form part of a stem and characters &ldquo;<tt>()<>[]{}</tt>&rdquo; to
            define stem regions and pseudoknots.</p>
        <p><tt>-t</tt><br>Specify a user starting tree file name in Newick format.</p>
        <p><tt>-T</tt><br>PTHREADS VERSION ONLY! Specify the number of threads you want to run.<br>
            Make sure to set &ldquo;<tt>-T</tt>&rdquo; to at most the number of CPUs you have on your machine,
            otherwise, there will be a huge performance decrease!</p>
        <p><tt>-u</tt><br>use the median for the discrete approximation of the GAMMA model of rate heterogeneity.<br>
            DEFAULT: OFF</p>
        <p><tt>-U</tt><br>Try to save memory by using SEV-based implementation for gap columns on large gappy
            alignments.<br>
            The technique is described here: <a href="http://www.biomedcentral.com/1471-2105/12/470">http://www.biomedcentral.com/1471-2105/12/470</a><br>
            This will only work for DNA and/or PROTEIN data and only with the SSE3 or AVX-vextorized version of the
            code.</p>
        <p><tt>-v</tt><br>Display version information.</p>
        <p><tt>-V</tt><br>Disable rate heterogeneity among sites model and use one without rate heterogeneity
            instead.<br>
            Only works if you specify the CAT model of rate heterogeneity.<br>
            DEFAULT: use rate heterogeneity
        <p><tt>-w</tt><br>FULL (!) path to the directory into which RAxML shall write its output files.<br>
            DEFAULT: current directory</p>
        <p><tt>-W</tt><br>Sliding window size for leave-one-out site-specific placement bias algorithm.<br>
            Only effective when used in combination with &ldquo;<tt>-f&nbsp;S</tt>&rdquo;<br>
            DEFAULT: 100 sites</p>
        <p><tt>-x</tt><br>Specify an integer number (random seed) and turn on rapid bootstrapping.<br>
            CAUTION: unlike in version 7.0.4 RAxML will conduct rapid BS replicates under the model of rate
            heterogeneity you specified via &ldquo;<tt>-m</tt>&rdquo; and not by default under <tt>CAT</tt>.</p>
        <p><tt>-X</tt><br>
            Same as the &ldquo;<tt>-y</tt>&rdquo; option below, however the parsimony search is more superficial.<br>
            RAxML will only do a randomized stepwise addition order parsimony tree reconstruction without performing any
            additional SPRs.<br>
            This may be helpful for very broad whole-genome datasets, since this can generate topologically more
            different starting trees.<br>
            DEFAULT: OFF</p>
        <p><tt>-y</tt><br>If you want to only compute a parsimony starting tree with RAxML specify &ldquo;<tt>-y</tt>&rdquo;,
            the program will exit after computation of the starting tree.<br>
            DEFAULT: OFF</p>
        <p><tt>-Y</tt><br>Pass a quartet grouping file name defining four groups from which to draw quartets The file
            input format must contain 4 groups in the following form:<br>
            <tt>(Chicken, Human, Loach), (Cow, Carp), (Mouse, Rat, Seal), (Whale, Frog);</tt><br>
            Only works in combination with <tt>-f&nbsp;q</tt>!</p>
        <p><tt>-z</tt><br>Specify the file name of a file containing multiple trees e.&nbsp;g. from a bootstrap that
            shall be used to draw bipartition values onto a tree provided with &ldquo;<tt>-t</tt>&rdquo;.<br>
            It can also be used to compute per site log likelihoods in combination with &ldquo;<tt>-f&nbsp;g</tt>&rdquo;
            and to read a bunch of trees for a couple of other options (&ldquo;<tt>-f&nbsp;h</tt>&rdquo;, &ldquo;<tt>-f&nbsp;m</tt>&rdquo;, &ldquo;<tt>-f&nbsp;n</tt>&rdquo;).
        </p>
        <p><tt>-#</tt>|<tt>-N</tt><br>Specify the number of alternative runs on distinct starting trees.<br>
            In combination with the &ldquo;<tt>-b</tt>&rdquo; option, this will invoke a multiple boostrap analysis.<br>
            Note that &ldquo;<tt>-N</tt>&rdquo; has been added as an alternative since &ldquo;<tt>-#</tt>&rdquo;
            sometimes caused problems with certain MPI job submission systems, since &ldquo;<tt>-#</tt>&rdquo; is often
            used to start comments.<br>
            If you want to use the bootstopping criteria specify &ldquo;<tt>-#&nbsp;autoMR</tt>&rdquo; or &ldquo;<tt>-#&nbsp;autoMRE</tt>&rdquo;
            or &ldquo;<tt>-#&nbsp;autoMRE_IGN</tt>&rdquo; for the majority-rule tree based criteria (see <tt>-I</tt>
            option) or &ldquo;<tt>-#&nbsp;autoFC</tt>&rdquo; for the frequency-based criterion.<br>
            Bootstopping will only work in combination with &ldquo;<tt>-x</tt>&rdquo; or &ldquo;<tt>-b</tt>&rdquo;<br>
            DEFAULT: 1 single analysis</p>
        <p><tt>--mesquite</tt><br>
            Print output files that can be parsed by Mesquite.<br>
            DEFAULT: Off<br>
        <p><tt>--silent</tt><br>
            Disables printout of warnings related to identical sequences and entirely undetermined sites in the
            alignment.<br>
            DEFAULT: Off</p>
        <p><tt>--no-seq-check</tt><br>
            Disables checking the input MSA for identical sequences and entirely undetermined sites.<br>
            Enabling this option may save time, in particular for large phylogenomic alignments.<br>
            Before using this, make sure to check the alignment using the &ldquo;<tt>-f&nbsp;c</tt>&rdquo; option!<br>
            DEFAULT: Off</p>
        <p><tt>--no-bfgs</tt><br>
            Disables automatic usage of BFGS method to optimize GTR rates on unpartitioned DNA datasets.<br>
            DEFAULT: BFGS on</p>
        <p><tt>--asc-corr</tt><br>
            Allows to specify the type of ascertainment bias correction you wish to use. There are three types
            available:<br>
            <tt>--asc-corr=lewis</tt>: the standard correction by Paul Lewis.<br>
            <tt>--asc-corr=felsenstein</tt>: a correction introduced by Joe Felsenstein that allows to explicitely
            specify the number of invariable sites (if known) one wants to correct for.<br>
            <tt>--asc-corr=stamatakis</tt>: a correction introduced by myself that allows to explicitely specify the
            number of invariable sites for each character (if known) one wants to correct for.<br>
            For further details please refer to the manual!</p>
        <p><tt>--flag-check</tt><br>
            When using this option, RAxML will only check if all command line flags specifed are available and then exit
            with a message listing all invalid command line flags or with a message stating that all flags are valid.
        </p>
        <p><tt>--auto-prot=ml|bic|aic|aicc</tt><br>
            When using automatic protein model selection you can chose the criterion for selecting these models.<br>
            RAxML will test all available prot subst. models except for <tt>LG4M</tt>, <tt>LG4X</tt> and <tt>GTR</tt>-based
            models, with and without empirical base frequencies.<br>
            You can chose between ML score based selection and the BIC, AIC, and AICc criteria.<br>
            DEFAULT: ml</p>
        <p><tt>--epa-keep-placements=number</tt><br>
            specify the number of potential placements you want to keep for each read in the EPA algorithm.<br>
            Note that, the actual values printed will also depend on the settings for
            <tt>--epa-prob-threshold=threshold</tt>!<br>
            DEFAULT: 7</p>
        <p><tt>--epa-prob-threshold=threshold</tt>
            specify a percent threshold for including potential placements of a read depending on the maximum placement
            weight for this read. If you set this value to 0.01 placements that have a placement weight of 1 per cent of
            the maximum placement will still be printed to file if the setting of <tt>--epa-keep-placements</tt> allows
            for it.<br>
            DEFAULT: 0.01</p>
        <p><tt>--epa-accumulated-threshold=threshold</tt><br>
            specify an accumulated likelihood weight threshold for which different placements of read are printed to
            file. Placements for a read will be printed until the sum of their placement weights has reached the
            threshold value.<br>
            Note that, this option can neither be used in combination with <tt>--epa-prob-threshold</tt> nor with <tt>--epa-keep-placements</tt>!
        </p>
        <p><tt>--JC69</tt><br>
            specify that all DNA partitions will evolve under the Jukes-Cantor model, this overrides all other model
            specifications for DNA partitions.<br>
            DEFAULT: Off</p>
        <p><tt>--K80</tt><br>
            specify that all DNA partitions will evolve under the K80 model, this overrides all other model
            specifications for DNA partitions.<br>
            DEFAULT: Off</p>
        <p><tt>--asc-miss=fraction</tt><br>
            specify the fraction of missing data in the variable sites of the alignment you are trying to correct for
            ascertainment bias.<br>
            Experimental option, needs to be tested!</p>
    </div>
    </body></html>
